pkg/ingester/circuitbreaker.go

```diff
@@ -245,11 +237,87 @@ func (cb *circuitBreaker) recordResult(errs ...error) error {
 	for _, err := range errs {
 		if err != nil && isCircuitBreakerFailure(err) {
 			cb.cb.RecordFailure()
-			cb.metrics.circuitBreakerResults.WithLabelValues(circuitBreakerResultError).Inc()
+			cb.metrics.circuitBreakerResults.WithLabelValues(cb.requestType, circuitBreakerResultError).Inc()
 			return err
 		}
 	}
 	cb.cb.RecordSuccess()
-	cb.metrics.circuitBreakerResults.WithLabelValues(circuitBreakerResultSuccess).Inc()
+	cb.metrics.circuitBreakerResults.WithLabelValues(cb.requestType, circuitBreakerResultSuccess).Inc()
 	return nil
 }
+
+type ingesterCircuitBreaker struct {
+	push *circuitBreaker
+	read *circuitBreaker
+}
+
+func newIngesterCircuitBreaker(pushCfg CircuitBreakerConfig, readCfg CircuitBreakerConfig, logger log.Logger, registerer prometheus.Registerer) *ingesterCircuitBreaker {
+	prCB := &ingesterCircuitBreaker{}
+	state := func(requestType string) circuitbreaker.State {
+		switch requestType {
+		case circuitBreakerPushRequestType:
+			if prCB.push.isActive() {
+				return prCB.push.cb.State()
+			}
+		case circuitBreakerReadRequestType:
+			if prCB.read.isActive() {
+				return prCB.read.cb.State()
+			}
+		}
+		return -1
+	}
+	metrics := newCircuitBreakerMetrics(registerer, state, []string{circuitBreakerPushRequestType, circuitBreakerReadRequestType})
+	prCB.push = newCircuitBreaker(pushCfg, metrics, circuitBreakerPushRequestType, logger)
+	prCB.read = newCircuitBreaker(readCfg, metrics, circuitBreakerReadRequestType, logger)
+	return prCB
+}
+
+func (cb *ingesterCircuitBreaker) activate() {
+	if cb == nil {
+		return
+	}
+	cb.push.activate()
+	cb.read.activate()
+}
+
+// tryPushAcquirePermit tries to acquire a permit to use the push circuit breaker and returns whether a permit was acquired.
+// If it was possible, tryPushAcquirePermit returns a function that should be called to release the acquired permit.
+// If it was not possible, the causing error is returned.
+func (cb *ingesterCircuitBreaker) tryPushAcquirePermit() (func(time.Duration, error), error) {
+	if cb == nil || !cb.push.isActive() {
+		return func(time.Duration, error) {}, nil
+	}
+
+	pushAcquiredPermit, err := cb.push.tryAcquirePermit()
+	if err != nil {
+		return nil, err
+	}
+	return func(duration time.Duration, err error) {
+		if pushAcquiredPermit {
+			_ = cb.push.finishRequest(duration, cb.push.cfg.RequestTimeout, err)
```

Comment:
Why do we need to pass to `cb.push.foo()` a config from `cb.push.cfg`?
I think we can refactor `finishRequest` to read it's own config.

Reply:
A good catch. Thank you